sdk/src/dlob/DLOB.ts

```diff
@@ -0,0 +1,863 @@
+						slot
+					);
+
+					return bestBidPrice.gt(currentBidPrice)
+						? bestGenerator
+						: currentGenerator;
+				}
+			);
+
+			if (!bestGenerator.next.done) {
+				yield bestGenerator.next.value;
+				bestGenerator.next = bestGenerator.generator.next();
+			} else {
+				bidsExhausted = true;
+			}
+		}
+	}
+
+	findCrossingOrders(
+		askNode: DLOBNode,
+		bidNode: DLOBNode,
+		oraclePriceData: OraclePriceData,
+		slot: number
+	): {
+		crossingNodes?: NodeToFill;
+		exhaustedSide?: Side;
+	} {
+		const bidPrice = bidNode.getPrice(oraclePriceData, slot);
+		const askPrice = askNode.getPrice(oraclePriceData, slot);
+
+		// contains market order
+		const containsMarketOrder =
+			(askNode.order && isOneOfVariant(askNode.order.orderType, ['market'])) ||
+			(bidNode.order && isOneOfVariant(bidNode.order.orderType, ['market']));
+		if (!containsMarketOrder && bidPrice.lt(askPrice)) {
+			if (askNode.isVammNode() && !bidNode.isVammNode()) {
+				return {
+					exhaustedSide: 'bid',
+				};
+			} else if (!askNode.isVammNode() && bidNode.isVammNode()) {
+				return {
+					exhaustedSide: 'ask',
+				};
+			} else {
+				return {
+					exhaustedSide: 'both',
+				};
+			}
+		}
+
+		const bidOrder = bidNode.order;
+		const askOrder = askNode.order;
+
+		// User bid crosses the vamm ask
+		// Cant match orders
+		if (askNode.isVammNode()) {
+			if (!isAuctionComplete(bidOrder, slot)) {
+				return {
+					exhaustedSide: 'bid',
+				};
+			}
+			return {
+				crossingNodes: {
+					node: bidNode,
+				},
+				exhaustedSide: 'bid',
+			};
+		}
+
+		// User ask crosses the vamm bid
+		// Cant match orders
+		if (bidNode.isVammNode()) {
+			if (!isAuctionComplete(askOrder, slot)) {
+				return {
+					exhaustedSide: 'ask',
+				};
+			}
+			return {
+				crossingNodes: {
+					node: askNode,
+				},
+				exhaustedSide: 'ask',
+			};
+		}
+
+		const bidBaseRemaining = bidOrder.baseAssetAmount.sub(
+			bidOrder.baseAssetAmountFilled
+		);
+		const askBaseRemaining = askOrder.baseAssetAmount.sub(
+			askOrder.baseAssetAmountFilled
+		);
+
+		let exhaustedSide: Side;
+		if (bidBaseRemaining.eq(askBaseRemaining)) {
+			exhaustedSide = 'both';
+		} else if (bidBaseRemaining.gt(askBaseRemaining)) {
+			exhaustedSide = 'ask';
+		} else {
+			exhaustedSide = 'bid';
+		}
+
+		// update the orders as if they fill
+		if (exhaustedSide === 'ask') {
+			bidNode.order.baseAssetAmountFilled =
+				bidOrder.baseAssetAmountFilled.add(askBaseRemaining);
+			askNode.order.baseAssetAmountFilled = askOrder.baseAssetAmount;
+		} else if (exhaustedSide === 'bid') {
+			askNode.order.baseAssetAmountFilled =
+				askOrder.baseAssetAmountFilled.add(bidBaseRemaining);
+			bidNode.order.baseAssetAmountFilled = bidOrder.baseAssetAmount;
+		} else {
+			askNode.order.baseAssetAmountFilled = askOrder.baseAssetAmount;
+			bidNode.order.baseAssetAmountFilled = bidOrder.baseAssetAmount;
+		}
+
+		// Two maker orders cross
+		if (bidOrder.postOnly && askOrder.postOnly) {
```

Comment:
this should be higher and should stop you from updating the baseAssetAmountFilled?